### DISCUSSION TITLE
2.3 序列

### DISCUSSION THREAD

**Discussion by giscus** [csfive/composing-programs-zh#108] [Announcements]
# 2.3 序列

SICP Python 中文翻译

https://composingprograms.netlify.app/2/3

**Comment by 258yujin** [comment]
树递归这里看晕了都

**Comment by sil3nceyang** [comment]
gg在树递归和链表递归这里

**Comment by ciron-nine** [comment]
我感觉这节排版就有点烂了，先不说链表、树的实现在于知道孩子结点的位置都没讲清楚。十分不建议没学过数据结构的同学们硬学。这里梗概讲的很模糊，而由于链表和数都是高度递归的结构，你用[]进行分层，十分的混乱。
就拿最简单的讲，至少要先讲链表再讲树啊，链表只有一个后续结点，而树则是有多个后续结点，按正常道理来讲，也是应该先讲清楚一个，再去讲多个啊。

**Reply by CharlieCRX** [reply]
这里学习起来实在是太困难了！让我感到极度难受的地方：
1. 列表推导式的一般形式概念；
2. `完美数`是等于其约数之和的正整数。
3. 树：列表中嵌套列表，树的数据抽象
4. 树递归（Tree-recursive）函数可用于构造树： `The nth Fibonacci tree`
5. 分割树`partition_tree(n, m)`，以及涉及的代码，还利用了序列抽象的切片定义了右二叉树`right_binarize(tree)`
6. 链表的抽象数据表示
7. `满足序列抽象`的条件：长度`len_link(s)`和元素选择`getitem_link(s, i)`
8. 重头递归构造：1.7节中的 `count_partitions`

**Reply by CharlieCRX** [reply]
其实我也在思考怎么给新人排版这边的内容更好。这里的材料中提供的背景知识基本都是架空的（你并没有对应的背景基础），然后再在对应的背景下实现代码逻辑。
本来提供的抽象概念就够让人思考一段时间了，随着概念而来的代码逻辑更是让阅读时候的我雪上加霜。

**Reply by jackydra317** [reply]
译者大大是否可能把“树”画个图出来呢？一堆[]符号，硬看还是知道大概什么意思，但确实头晕……

**Comment by TrigYei** [comment]
right_binarize(tree)部分代码好像有点问题。
 return [right_binarize(b) for b in tree] 这里的b可能是整数，而is_leaf(tree)要求tree是列表

**Reply by SwordKillCowardice** [reply]
其实没有呢，b最少也是[1]这样的列表套一个数字，因为Branch选择函数会在原先的分支前面再套一个[]，然后才返回

**Reply by AlanAhoy** [reply]
我也觉得好像有点问题，第一次递归调用right_binarize(b)的时候b是1，在branches(tree)函数里会报错TypeError: 'int' object is not subscriptable

**Reply by Ya-kei** [reply]
确实有问题，树的第一个元素是标号而不是一棵树，传进 right_binarize 的话类型不对，应该改成 `return [tree[0]] + [right_binarize(b) for b in branches(tree)]`

**Comment by shenQQQ** [comment]
翻译的真的好👍
数据结构更多是需求驱动，有了需求自然而然就得构造出这样的结构。但是有些相关的概念名词在中文教科书里都特别晦涩。大佬都翻译的很清晰，比如闭包这里好多中文材料里的说法都不一样，导致之前一直似懂非懂，这次真的彻底看懂了👍

**Comment by HealerOS** [comment]
作为过来人表示，这里的树和链表不用看，去看数据结构相关的课程或者书籍中对这两种数据结构的介绍会更容易理解

**Reply by sleepy-shawn** [reply]
好的谢谢，那就不纠结继续看下面的课程了

**Comment by jahox0102** [comment]
树和链表看晕了

**Comment by KKKKoishi514** [comment]
已经用c++实现过链表和二叉树了，结果这里讲的树我居然看不明白...
推荐找本数据结构的书看这个，这一节写的过于逆天

**Comment by MasterpieceXu** [comment]
这一节比较难，可以和ai对话让ai解释一下，树递归像是把递归的过程记录下来（用列表的方式）

**Comment by luxianche9** [comment]
链表和树直接跳过不看了, 学数据结构与算法的时候再看, 知道数据结构的实现与操作分离就行了

**Comment by wuchu576** [comment]
看了3天，勉强把树和链表看完，但还是一知半解，已经想要跳过。无意翻到评论，应该早点看评论的。

**Comment by lzxlhdj** [comment]
拼尽全力想理解，可惜1晚了看到评论，不过说真的，用[]来表示树真的看的令人头大

**Comment by Halle-0620** [comment]
救命……终于把树和链表啃完了……这个树简直是太抽象了，看到大家都觉得有难度好受一些……唉，我之后一定要先看看下面的评论我再接着看，这样我就晓得先去看别的教材这一部分的内容了QAQ

**Comment by mich-ui80** [comment]
这节课里用slicing生成二叉树形式的代码是不是有点问题呀，我检查了好几遍但是依然报错。错误原因好像是用到的is_leaf是基于tree设计的，但是传入的东西可能只是一个很普通的数组，而不是tree形式的数组，因此就会报错。重新写一个is_leaf函数应该可以解决。（我是新手，有大佬能解决一下嘛）

**Reply by mich-ui80** [reply]
链表的partition那里好像也有一点问题

**Comment by CYXC11** [comment]
这里的链表跟c的广义表很相似阿

**Reply by CYXC11** [reply]
认真看了看 感觉更像右分叉的二叉树 和2.3.6的末尾的代码实现对应上了

**Comment by AlanRosston** [comment]
讲得太抽象了...还是面向对象的写法容易懂些
